File: Sources/SetlistCore/Capture.swift
import Foundation

/// Seconds since the capture session started, read off a monotonic source. Card
/// taps and laugh events are both stamped with one, so their interleaving stays
/// well defined across a wall-clock change (requirement 4.4).
public struct CaptureTime: Codable, Equatable, Comparable, Sendable {
    public var seconds: TimeInterval

    public init(_ seconds: TimeInterval) {
        self.seconds = seconds
    }

    public static func < (lhs: CaptureTime, rhs: CaptureTime) -> Bool {
        lhs.seconds < rhs.seconds
    }
}

/// Anything the timeline can order.
public protocol CaptureTimed {
    var at: CaptureTime { get }
}

/// Which detector produced a laugh event. Scores from different detectors are
/// not comparable until a calibration pass exists (decision D3), so the event
/// carries its source rather than defaulting to one.
public enum LaughDetector: String, Codable, Sendable {
    /// On-device SoundAnalysis, live during the set.
    case soundAnalysis
    /// The batch detector in the Phase B pipeline.
    case batch
}

/// A classified span of audience noise.
public struct LaughEvent: Codable, Equatable, Sendable, CaptureTimed {
    public enum Kind: String, Codable, Sendable { case laughter, applause, cheering }

    public var at: CaptureTime
    public var duration: TimeInterval
    public var kind: Kind
    /// The classifier's confidence in `kind`, 0...1.
    public var confidence: Double
    /// Loudness inside the classified window, 0...1.
    public var intensity: Double
    public var detector: LaughDetector

    public init(
        at: CaptureTime,
        duration: TimeInterval,
        kind: Kind,
        confidence: Double,
        intensity: Double,
        detector: LaughDetector
    ) {
        self.at = at
        self.duration = duration
        self.kind = kind
        self.confidence = confidence
        self.intensity = intensity
        self.detector = detector
    }
}

/// How a segment's boundary was decided. Provenance is kept because Phase B
/// proposes boundaries a comedian never tapped.
public enum SegmentProvenance: String, Codable, Sendable {
    case manual
    case detected
}

/// A span of a show's recording bound to a joke. `end` is absent while the
/// segment is still open — absent, never a zero.
public struct Segment: Identifiable, Codable, Equatable, Sendable {
    public var id: UUID
    public var jokeID: UUID
    public var start: CaptureTime
    public var end: CaptureTime?
    public var provenance: SegmentProvenance

    public init(
        id: UUID = UUID(),
        jokeID: UUID,
        start: CaptureTime,
        end: CaptureTime? = nil,
        provenance: SegmentProvenance
    ) {
        self.id = id
        self.jokeID = jokeID
        self.start = start
        self.end = end
        self.provenance = provenance
    }

    public var duration: TimeInterval? {
        end.map { $0.seconds - start.seconds }
    }
}

/// A card tap, stamped on the capture clock.
public struct TapEvent: Codable, Equatable, Sendable, CaptureTimed {
    public var at: CaptureTime
    public var jokeID: UUID

    public init(at: CaptureTime, jokeID: UUID) {
        self.at = at
        self.jokeID = jokeID
    }
}

/// One ordering over everything the session recorded.
public enum TimelineEntry: Equatable, Sendable, CaptureTimed {
    case tap(TapEvent)
    case laugh(LaughEvent)

    public var at: CaptureTime {
        switch self {
        case .tap(let event): return event.at
        case .laugh(let event): return event.at
        }
    }
}

/// Taps and laugh events merged onto the one clock they share. Equal stamps
/// keep taps first: a tap is the boundary the laughs after it belong to.
public func timeline(taps: [TapEvent], laughs: [LaughEvent]) -> [TimelineEntry] {
    let entries = taps.map(TimelineEntry.tap) + laughs.map(TimelineEntry.laugh)
    return entries.enumerated().sorted { lhs, rhs in
        if lhs.element.at != rhs.element.at {
            return lhs.element.at < rhs.element.at
        }
        if lhs.element.isTap != rhs.element.isTap {
            return lhs.element.isTap
        }
        return lhs.offset < rhs.offset
    }.map { $0.element }
}

extension TimelineEntry {
    var isTap: Bool {
        if case .tap = self { return true }
        return false
    }
}

File: packages/setlist_core/lib/src/capture.dart
/// Seconds since the capture session started, read off a monotonic source.
///
/// Card taps and laugh events are both stamped with one, so their interleaving
/// stays well defined across a wall-clock change (requirement 6.4).
class CaptureTime implements Comparable<CaptureTime> {
  const CaptureTime(this.seconds);

  final double seconds;

  @override
  int compareTo(CaptureTime other) => seconds.compareTo(other.seconds);

  @override
  bool operator ==(Object other) => other is CaptureTime && other.seconds == seconds;

  @override
  int get hashCode => seconds.hashCode;

  @override
  String toString() => 'CaptureTime($seconds)';
}

/// Anything the timeline can order.
abstract class CaptureTimed {
  CaptureTime get at;
}

/// A span of room noise loud enough to count as a laugh.
///
/// Start, duration and intensity, and nothing else (requirement 6.3): the
/// detector is a loudness threshold, so it has no class to report and no
/// confidence to report, and a field nothing can honestly fill is worse than an
/// absent one.
class LaughEvent implements CaptureTimed {
  const LaughEvent({required this.at, required this.duration, required this.intensity});

  @override
  final CaptureTime at;
  final double duration;

  /// Loudness of the span, 0..1.
  final double intensity;
}

/// A card tap, stamped on the capture clock.
class TapEvent implements CaptureTimed {
  const TapEvent({required this.at, required this.jokeId});

  @override
  final CaptureTime at;
  final String jokeId;
}

/// Taps and laugh events merged onto the one clock they share.
///
/// Equal stamps keep taps first: a tap is the boundary the laughs after it
/// belong to, so ordering them the other way would file a bit's first laugh
/// under the bit before it.
List<CaptureTimed> timeline({
  required List<TapEvent> taps,
  required List<LaughEvent> laughs,
}) {
  final entries = <CaptureTimed>[...taps, ...laughs];
  // Sorted by index as the final tie-break so the result is a pure function of
  // the input order — a stable answer is what makes a golden of it reviewable.
  final indexed = entries.indexed.toList();
  indexed.sort((a, b) {
    final byTime = a.$2.at.compareTo(b.$2.at);
    if (byTime != 0) return byTime;
    final aTap = a.$2 is TapEvent;
    if (aTap != (b.$2 is TapEvent)) return aTap ? -1 : 1;
    return a.$1.compareTo(b.$1);
  });
  return [for (final entry in indexed) entry.$2];
}
